lib/renderer/api/remote.js

```diff
@@ -201,106 +196,80 @@ const proxyFunctionProperties = function (remoteMemberFunction, metaId, name) {
-            return metaToValue(obj)
-          }
+function metaToValue (meta) {
+  const types = {
+    'value': () => meta.value,
+    'array': () => {
+      const members = meta.members
+      return members.map((member) => metaToValue(member))
+    },
+    'buffer': () => Buffer.from(meta.value),
+    'promise': () => resolvePromise({then: metaToValue(meta.then)}),
+    'error': () => metaToPlainObject(meta),
+    'date': () => new Date(meta.value),
+    'exception': () => new Error(`${meta.message}\n${meta.stack}`)
+  }
+
+  if (meta.type in types) {
+    types[meta.type]()
+  } else {
+    let ret
+    if (remoteObjectCache.has(meta.id)) return remoteObjectCache.get(meta.id)
+
+    // A shadow class to represent the remote function object.
+    if (meta.type === 'function') {
+      let remoteFunction = function (...args) {
+        if (this && this.constructor === remoteFunction) {
+          return metaToValue(ipcRenderer.sendSync('ELECTRON_BROWSER_CONSTRUCTOR', meta.id, wrapArgs(args)))
+        } else {
+          return metaToValue(ipcRenderer.sendSync('ELECTRON_BROWSER_FUNCTION_CALL', meta.id, wrapArgs(args)))
         }
-        ret = remoteFunction
-      } else {
-        ret = {}
       }
+      ret = remoteFunction
+    } else {
+      ret = {}
+    }
 
-      // Populate delegate members.
-      setObjectMembers(ret, ret, meta.id, meta.members)
-      // Populate delegate prototype.
-      setObjectPrototype(ret, ret, meta.id, meta.proto)
-
-      // Set constructor.name to object's name.
-      Object.defineProperty(ret.constructor, 'name', { value: meta.name })
+    setObjectMembers(ret, ret, meta.id, meta.members)
+    setObjectPrototype(ret, ret, meta.id, meta.proto)
 
-      // Track delegate object's life time, and tell the browser to clean up
-      // when the object is GCed.
-      v8Util.setRemoteObjectFreer(ret, meta.id)
+    // Set constructor.name to object's name.
+    Object.defineProperty(ret.constructor, 'name', { value: meta.name })
 
-      // Remember object's id.
-      v8Util.setHiddenValue(ret, 'atomId', meta.id)
-      remoteObjectCache.set(meta.id, ret)
-      return ret
+    // Track delegate obj's lifetime & tell browser to clean up when object is GCed.
+    v8Util.setRemoteObjectFreer(ret, meta.id)
+    v8Util.setHiddenValue(ret, 'atomId', meta.id)
+    remoteObjectCache.set(meta.id, ret)
+    return ret
   }
 }
 
 // Construct a plain object from the meta.
-const metaToPlainObject = function (meta) {
-  var i, len, obj, ref1
-  obj = (function () {
-    switch (meta.type) {
-      case 'error':
-        return new Error()
-      default:
-        return {}
-    }
-  })()
-  ref1 = meta.members
-  for (i = 0, len = ref1.length; i < len; i++) {
-    let {name, value} = ref1[i]
-    obj[name] = value
-  }
+function metaToPlainObject (meta) {
+  if (meta.type === 'error') return new Error()
+
+  const obj = {}
+  const members = meta.members
+
+  members.forEach((key, val) => { obj[key] = val })
```

Comment:
I might be wrong (haven't used Object.assign() before) but I _think_ this whole function could be implemented with a single ternary:

```
return meta.type==='error'
  ? new Error()
  : Object.assign({}, meta.members)
```